lightning/src/onion_message/messenger.rs

```diff
@@ -243,6 +244,86 @@ impl OnionMessageRecipient {
 	}
 }
 
+/// A enum to handle received [`OnionMessage`]
+pub enum ReceivedOnionMessage<R, T>
+where
+	R: RespondFunction<T>,
+	T: OnionMessageContents
+{
+	WithReplyPath {
+		message: T,
+		path_id: Option<[u8; 32]>,
+		responder: Responder<R, T>,
```

Comment:
Hmmm... it's not really a reply but rather a message that can be replied to. Regardless of naming, I'm not sure I see how such an interface would be any better or work for that matter. The handler would need a reference to `OnionMessenger`, which would result in a circular reference since `OnionMessenger` already has a reference to the handler. Am I misunderstanding what you're proposing?

Reply:
@shaavan Discussed further offline with @TheBlueMatt. Seems we'll run into the circular reference and some other lifetime issues in the async case. We'd soon see this when writing a test where the handler tries to store the `ReceivedOnionMessage` for later processing.

So we'll have to drop the `ResponseFunction` unfortunately. Instead, we'll need to do something like the following:
- Have `Responder::respond` return some sort of `ResponseInstructions<T>` (exact name TBD... maybe `OnionMessageResponse` instead?) which would include the response message and (in a follow-up) a way to indicated that a reply path should be built to include with the response
- Change the handlers back to have a return value only this time using type `ResponseInstructions<T>` where `T` is the concrete message type for the handler.
- Instead of using an `Option` in the handler's return type you could possibly make `ResponseInstructions` an enum with a `NoResponse` variant, FWIW.
- When OnionMessenger gets the returned `ResponseInstructions` it would respond appropriately.

That way a handler can take ownership of `ReceivedOnionMessage` and produce a `ResponseInstructions` asynchronously. In another follow-up PR we would add a method to `OnionMessenger` that takes a `ResponseInstructions` and responds just as it does in the sync case.